java/dataset/src/test/java/org/apache/arrow/dataset/file/TestDatasetFileWriter.java

```diff
@@ -0,0 +1,140 @@
+    ScanOptions options = new ScanOptions(new String[0], 100);
+    final File writtenFolder = TMP.newFolder();
+    final String writtenParquet = writtenFolder.toURI().toString();
+
+    try (FileSystemDatasetFactory factory = new FileSystemDatasetFactory(rootAllocator(), NativeMemoryPool.getDefault(),
+        FileFormat.PARQUET, sampleParquet);
+         final Dataset dataset = factory.finish();
+         final Scanner scanner = dataset.newScan(options);
+         final ArrowScannerReader reader = new ArrowScannerReader(scanner, rootAllocator());
+    ) {
+      DatasetFileWriter.write(rootAllocator(), reader, FileFormat.PARQUET, writtenParquet, new String[]{"id", "name"}, 100, "data_{i}");
+      final Set<String> expectedOutputFiles = new HashSet<>(
+          Arrays.asList("id=1/name=a/data_0", "id=2/name=b/data_0", "id=3/name=c/data_0", "id=2/name=d/data_0"));
+      final Set<String> outputFiles = FileUtils.listFiles(writtenFolder, null, true)
+          .stream()
+          .map(file -> {
+            return writtenFolder.toURI().relativize(file.toURI()).toString();
+          })
+          .collect(Collectors.toSet());
+      Assert.assertEquals(expectedOutputFiles, outputFiles);
+    }
+  }
+
+  private void assertParquetFileEquals(String expectedURI, String actualURI) throws Exception {
+    final FileSystemDatasetFactory expectedFactory = new FileSystemDatasetFactory(
+        rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.PARQUET, expectedURI);
+    final FileSystemDatasetFactory actualFactory = new FileSystemDatasetFactory(
+        rootAllocator(), NativeMemoryPool.getDefault(), FileFormat.PARQUET, actualURI);
+    List<ArrowRecordBatch> expectedBatches = collectResultFromFactory(expectedFactory,
+        new ScanOptions(new String[0], 100));
+    List<ArrowRecordBatch> actualBatches = collectResultFromFactory(actualFactory,
+        new ScanOptions(new String[0], 100));
+    try (
+        VectorSchemaRoot expectVsr = VectorSchemaRoot.create(expectedFactory.inspect(), rootAllocator());
+        VectorSchemaRoot actualVsr = VectorSchemaRoot.create(actualFactory.inspect(), rootAllocator())) {
+
+      // fast-fail by comparing metadata
+      Assert.assertEquals(expectedBatches.toString(), actualBatches.toString());
+      // compare ArrowRecordBatches
+      Assert.assertEquals(expectedBatches.size(), actualBatches.size());
+      VectorLoader expectLoader = new VectorLoader(expectVsr);
+      VectorLoader actualLoader = new VectorLoader(actualVsr);
+      for (int i = 0; i < expectedBatches.size(); i++) {
+        expectLoader.load(expectedBatches.get(i));
+        actualLoader.load(actualBatches.get(i));
+        for (int j = 0; j < expectVsr.getFieldVectors().size(); j++) {
+          FieldVector vector = expectVsr.getFieldVectors().get(i);
+          FieldVector otherVector = actualVsr.getFieldVectors().get(i);
+          Assert.assertTrue(VectorEqualsVisitor.vectorEquals(vector, otherVector));
```

Comment:
VectorSchemaRoot#equals should do this for you

Reply:
I try to use `VectorSchemaRoot#equals` to compare, but the result return false by [this issue](https://issues.apache.org/jira/browse/ARROW-18140) when comparing the schema info.

Reply:
Thanks for filing that. Can you add a comment like `// TODO(ARROW-18140): ...`?

Reply:
Already added the comment.